zkevm-circuits/src/evm_circuit/execution/error_oog_constant.rs

```diff
@@ -27,42 +34,94 @@ impl<F: Field> ExecutionGadget<F> for ErrorOOGConstantGadget<F> {
 
     fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
         let opcode = cb.query_cell();
+        cb.opcode_lookup(opcode.expr(), 1.expr());
+
         let gas_required = cb.query_cell();
 
+        cb.constant_gas_lookup(opcode.expr(), gas_required.expr());
         // Check if the amount of gas available is less than the amount of gas
         // required
         let insufficient_gas =
-            RangeCheckGadget::construct(cb, gas_required.expr() - cb.curr.state.gas_left.expr());
+            LtGadget::construct(cb, cb.curr.state.gas_left.expr(), gas_required.expr());
+        cb.require_equal(
+            "constant gas left is less than gas required ",
+            insufficient_gas.expr(),
+            1.expr(),
+        );
+
+        // current call must be failed.
+        cb.call_context_lookup(false.expr(), None, CallContextFieldTag::IsSuccess, 0.expr());
+        cb.call_context_lookup(
+            false.expr(),
+            None,
+            CallContextFieldTag::IsPersistent,
+            0.expr(),
+        );
 
-        // TODO: Handle root & internal call constraints and use ContextSwitchGadget to
-        // switch call context to caller's and consume all gas_left
+        // Go to EndTx only when is_root
+        let is_to_end_tx = cb.next.execution_state_selector([ExecutionState::EndTx]);
+        cb.require_equal(
+            "Go to EndTx only when is_root",
+            cb.curr.state.is_root.expr(),
+            is_to_end_tx,
+        );
+
+        // When it's a root call
+        cb.condition(cb.curr.state.is_root.expr(), |cb| {
+            // Do step state transition
+            cb.require_step_state_transition(StepStateTransition {
+                call_id: Same,
+                rw_counter: Delta(2.expr() + cb.curr.state.reversible_write_counter.expr()),
```

Comment:
I realize `handle_reversion` will add `reverted` operations which increase the `rw_counter`, increasing count is `reversible_write_counter`,  will update it in the spec.

Reply:
spec update https://github.com/privacy-scaling-explorations/zkevm-specs/pull/280 here.